rest/config.go

```diff
@@ -801,41 +792,64 @@ func (self *ServerConfig) MergeWith(other *ServerConfig) error {
 		if self.Databases[name] != nil {
 			return base.RedactErrorf("Database %q already specified earlier", base.UD(name))
 		}
+		if self.Databases == nil {
+			self.Databases = make(DbConfigMap)
+		}
 		self.Databases[name] = db
 	}
 	return nil
 }
 
-// Reads the command line flags and the optional config file.
-func ParseCommandLine() (err error) {
-	addr := flag.String("interface", DefaultInterface, "Address to bind to")
-	authAddr := flag.String("adminInterface", DefaultAdminInterface, "Address to bind admin interface to")
-	profAddr := flag.String("profileInterface", "", "Address to bind profile interface to")
-	configServer := flag.String("configServer", "", "URL of server that can return database configs")
-	deploymentID := flag.String("deploymentID", "", "Customer/project identifier for stats reporting")
-	couchbaseURL := flag.String("url", DefaultServer, "Address of Couchbase server")
-	poolName := flag.String("pool", DefaultPool, "Name of pool")
-	bucketName := flag.String("bucket", "sync_gateway", "Name of bucket")
-	dbName := flag.String("dbname", "", "Name of Couchbase Server database (defaults to name of bucket)")
-	pretty := flag.Bool("pretty", false, "Pretty-print JSON responses")
-	verbose := flag.Bool("verbose", false, "Log more info about requests")
-	logKeys := flag.String("log", "", "Log keys, comma separated")
-	logFilePath := flag.String("logFilePath", "", "Path to log files")
-	certpath := flag.String("certpath", "", "Client certificate path")
-	cacertpath := flag.String("cacertpath", "", "Root CA certificate path")
-	keypath := flag.String("keypath", "", "Client certificate key path")
+var (
+	addr, authAddr, profAddr, configServer, deploymentID, couchbaseURL, poolName, bucketName, dbName,
+	logKeys, logFilePath, certpath, cacertpath, keypath, defaultLogFilePathFlag *string
+)
```

Comment:
Better to avoid global variables here.  To support testability, split ParseCommandLine into ParseCommandLineArgs, and then 'InitializeFromCommandLine', where the latter takes the args as function parameters.  Can then unit test invoke InitializeFromCommandLine to provide test coverage.

Reply:
Removed the use of global variables; Revised the signature of `ParseCommandLine` function to `func ParseCommandLine(args []string) (err error) {...}` to accept the command-line parameters; then inside this call `flag.CommandLine.Parse(args)` instead of `flag.Parse()`. The resolved value of args would be `os.Args[1:]' when running sync gateway and we can pass custom parameters from tests for validating multiple scenarios. Also we can reset the value of flag in the tests like this:
`flag.CommandLine = flag.NewFlagSet(os.Args[0], flag.ExitOnError)`